src/main/java/com/atlassian/db/replica/api/KeyBasedThrottledCache.java

```diff
@@ -0,0 +1,43 @@
+package com.atlassian.db.replica.api;
+
+import com.atlassian.db.replica.spi.SuppliedCache;
+
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+
+/**
+ * Every read can potentially refresh the cache, however it can be loaded only by one thread at the time.
+ * Other threads instead of waiting are getting last known value.
+ */
+public class KeyBasedThrottledCache implements SuppliedCache<Long> {
```

Comment:
maybe it should be named in some simpler way like: KeyBasedNoConcurrentRefreshCache? WDYT?

Reply:
The loaded name is caused by having two responsibilities: keying per String + locking.

I infer the `String key` will represent a replica JDBC url or a tenant id. We already have a cross-replica and cross-tenant in-memory state: the `AbstractConnectionPoolProvider`. It can hold the `ConcurrentHashMap<String, SuppliedCache>`.
And the `SuppliedCache` itself would become just a `ReentrantLock` + a value field.

Alternatively, the `ConcurrentHashMap` aspect could be decoupled from the `if (lock.tryLock()` aspect as individual `SuppliedCache` impls.

Reply:
Let me gather all the consequences of `static` state:
* [testing and GC](https://github.com/atlassian-labs/db-replica/pull/68/files#r604083839)
* [limited to `Long`](https://github.com/atlassian-labs/db-replica/pull/68/files#r604068028)
* [magical implicit interactions](https://github.com/atlassian-labs/db-replica/pull/68/files#r604086184)
* [two responsibilities](https://github.com/atlassian-labs/db-replica/pull/68/files#r604088954)

Each on its own might be a nit-pick. But together it seems worth trying to get rid of `static`. Or at least isolate it from the valuable throttling feature.

Reply:
I also really like the idea of extracting key functionality and keeping only lock + value field here. Really nice hint Maciej!